File: concepts/preview-parser.ts
type Token =
  | { kind: "identifier"; value: string }
  | { kind: "int" }
  | { kind: "void" }
  | { kind: "return" }
  | { kind: "constant"; value: number }
  | { kind: "openParenthesis" }
  | { kind: "closeParenthesis" }
  | { kind: "openBrace" }
  | { kind: "closeBrace" }
  | { kind: "semicolon" };

type Program = { function: FunctionNode };
type FunctionNode = { identifier: string; statement: ReturnNode };
type ReturnNode = { expression: ConstantNode };
type ConstantNode = { value: number };

const patterns: [RegExp, (text: string) => Token][] = [
  [/^int\b/, () => ({ kind: "int" })],
  [/^void\b/, () => ({ kind: "void" })],
  [/^return\b/, () => ({ kind: "return" })],
  [/^[a-zA-Z_]\w*\b/, (text) => ({ kind: "identifier", value: text })],
  [/^[0-9]+\b/, (text) => ({ kind: "constant", value: Number(text) })],
  [/^\(/, () => ({ kind: "openParenthesis" })],
  [/^\)/, () => ({ kind: "closeParenthesis" })],
  [/^{/, () => ({ kind: "openBrace" })],
  [/^}/, () => ({ kind: "closeBrace" })],
  [/^;/, () => ({ kind: "semicolon" })],
];

function lex(input: string): Token[] {
  const tokens: Token[] = [];

  while (input !== "") {
    input = input.trimStart();
    if (input === "") break;

    const match = patterns.find(([pattern]) => pattern.test(input));
    if (!match) throw new Error(`unexpected character: '${input[0]}'`);

    const [pattern, createToken] = match;
    const text = pattern.exec(input)![0];
    tokens.push(createToken(text));
    input = input.slice(text.length);
  }

  return tokens;
}

class Parser {
  tokens: Token[];
  current = 0;

  constructor(tokens: Token[]) {
    this.tokens = tokens;
  }

  parseProgram(): Program {
    const program = { function: this.parseFunction() };
    if (this.current !== this.tokens.length) {
      throw new Error("unexpected trailing token");
    }
    return program;
  }

  parseFunction(): FunctionNode {
    this.expect("int");
    const identifier = this.parseIdentifier();
    this.expect("openParenthesis");
    this.expect("void");
    this.expect("closeParenthesis");
    this.expect("openBrace");
    const statement = this.parseStatement();
    this.expect("closeBrace");
    return { identifier, statement };
  }

  parseStatement(): ReturnNode {
    this.expect("return");
    const expression = this.parseExpression();
    this.expect("semicolon");
    return { expression };
  }

  parseExpression(): ConstantNode {
    return { value: this.parseInteger() };
  }

  parseIdentifier(): string {
    const token = this.take();
    if (token.kind !== "identifier") throw new Error("expected identifier");
    return token.value;
  }

  parseInteger(): number {
    const token = this.take();
    if (token.kind !== "constant") throw new Error("expected integer");
    return token.value;
  }

  expect(expected: Token["kind"]): void {
    if (this.take().kind !== expected) throw new Error("unexpected token");
  }

  take(): Token {
    const token = this.tokens[this.current];
    if (!token) throw new Error("unexpected end of input");
    this.current++;
    return token;
  }
}

const program = new Parser(lex("int main(void) { return 2; }")).parseProgram();
if (program.function.identifier !== "main" || program.function.statement.expression.value !== 2) {
  throw new Error("unexpected syntax tree");
}
console.log(
  `Program { Function = Function { Identifier = ${program.function.identifier}, Statement = Return { Expression = Constant { Value = ${program.function.statement.expression.value} } } } }`,
);

File: concepts/preview-parser.py
import re
from dataclasses import dataclass
from enum import Enum, auto
from typing import Optional, Union


class TokenKind(Enum):
    IDENTIFIER = auto()
    INT = auto()
    VOID = auto()
    RETURN = auto()
    CONSTANT = auto()
    OPEN_PARENTHESIS = auto()
    CLOSE_PARENTHESIS = auto()
    OPEN_BRACE = auto()
    CLOSE_BRACE = auto()
    SEMICOLON = auto()


@dataclass(frozen=True)
class Token:
    kind: TokenKind
    value: Optional[Union[str, int]] = None


PATTERNS = [
    (TokenKind.INT, re.compile(r"int\b")),
    (TokenKind.VOID, re.compile(r"void\b")),
    (TokenKind.RETURN, re.compile(r"return\b")),
    (TokenKind.IDENTIFIER, re.compile(r"[a-zA-Z_]\w*\b")),
    (TokenKind.CONSTANT, re.compile(r"[0-9]+\b")),
    (TokenKind.OPEN_PARENTHESIS, re.compile(r"\(")),
    (TokenKind.CLOSE_PARENTHESIS, re.compile(r"\)")),
    (TokenKind.OPEN_BRACE, re.compile(r"{")),
    (TokenKind.CLOSE_BRACE, re.compile(r"}")),
    (TokenKind.SEMICOLON, re.compile(r";")),
]


@dataclass(frozen=True)
class Program:
    function: "Function"


@dataclass(frozen=True)
class Function:
    identifier: str
    statement: "Return"


@dataclass(frozen=True)
class Return:
    expression: "Constant"


@dataclass(frozen=True)
class Constant:
    value: int


def lex(source: str) -> list[Token]:
    tokens = []
    while source:
        source = source.lstrip()
        if not source:
            break

        for kind, pattern in PATTERNS:
            match = pattern.match(source)
            if match:
                text = match.group()
                value = int(text) if kind is TokenKind.CONSTANT else text if kind is TokenKind.IDENTIFIER else None
                tokens.append(Token(kind, value))
                source = source[len(text):]
                break
        else:
            raise ValueError(f"unexpected character: {source[0]!r}")

    return tokens


class Parser:
    def __init__(self, tokens: list[Token]):
        self.tokens = tokens
        self.current = 0

    def parse_program(self) -> Program:
        program = Program(self.parse_function())
        if self.current != len(self.tokens):
            raise ValueError("unexpected trailing token")
        return program

    def parse_function(self) -> Function:
        self.expect(TokenKind.INT)
        identifier = self.parse_identifier()
        self.expect(TokenKind.OPEN_PARENTHESIS)
        self.expect(TokenKind.VOID)
        self.expect(TokenKind.CLOSE_PARENTHESIS)
        self.expect(TokenKind.OPEN_BRACE)
        statement = self.parse_statement()
        self.expect(TokenKind.CLOSE_BRACE)
        return Function(identifier, statement)

    def parse_statement(self) -> Return:
        self.expect(TokenKind.RETURN)
        expression = self.parse_expression()
        self.expect(TokenKind.SEMICOLON)
        return Return(expression)

    def parse_expression(self) -> Constant:
        return Constant(self.parse_integer())

    def parse_identifier(self) -> str:
        token = self.take()
        if token.kind is not TokenKind.IDENTIFIER or not isinstance(token.value, str):
            raise ValueError("expected identifier")
        return token.value

    def parse_integer(self) -> int:
        token = self.take()
        if token.kind is not TokenKind.CONSTANT or not isinstance(token.value, int):
            raise ValueError("expected integer")
        return token.value

    def expect(self, expected: TokenKind) -> None:
        if self.take().kind is not expected:
            raise ValueError("unexpected token")

    def take(self) -> Token:
        if self.current == len(self.tokens):
            raise ValueError("unexpected end of input")
        token = self.tokens[self.current]
        self.current += 1
        return token


program = Parser(lex("int main(void) { return 2; }")).parse_program()
assert program.function.identifier == "main"
assert program.function.statement.expression.value == 2
print(
    f"Program {{ Function = Function {{ Identifier = {program.function.identifier}, "
    f"Statement = Return {{ Expression = Constant {{ Value = {program.function.statement.expression.value} }} }} }} }}"
)
